### PR TITLE
Allow filtering variations by attributes in API.

### LINKED ISSUE #20349: /wc/v2/products/<product_id>/variations - filter by attribute/attribute_term
**Describe the bug**
Howdy!

When trying to use the REST API to display variations based on attribute, sorting by `attribute` and `attribute_term` consistently gives zero results. 

Documentation states that I should be able to filter the responses based on the these values: https://woocommerce.github.io/woocommerce-rest-api-docs/?shell#list-all-product-variations

Example: `/wp-json/wc/v2/products/125/variations/?attribute=pa_color&attribute_term=8`

This issue appears to only effect the`/products/{product_id}/variations/` endpoint, and the filter works correctly for the `/products/` endpoint.

Working Example:`/wp-json/wc/v2/products/attribute=pa_color&attribute_term=8`

I believe the cause is that the `$args` are built out in `prepare_objects_query()`, creating this (simplified) query:

```	
	$query  = new WP_Query();

	$query_args = array(
		'post_type'            => 'product_variation',
		'post_parent'            => '125',
		'tax_query'=>
			[
				[
					"taxonomy"=> "pa_color",
					"field"=> "term_id",
					"terms"=>'7',
				],
			],
	);
	
	$result = $query->query( $query_args );
```

The problem here is that product variations do not get assigned to a taxonomy; so technically the endpoint is accurate, however there won't be anything to display.

**To Reproduce**
1. Use the API on the `/product/` endpoint and filter it to ensure you have accurate values. Example:`/wp-json/wc/v2/products/attribute=pa_color&attribute_term=8`
2. Try the same filtering, except for product variations this time: `/wp-json/wc/v2/products/125/variations/?attribute=pa_color&attribute_term=8`

NOTE: if you have entered a bad attribute, you'll get every possible variation returned. If you enter correctly, you get nothing.

**Expected behavior**
If I was selling shirts, I'd expect to be able to use this endpoint for getting the details of a specific variation based on the supplied attributes.

Ideally, I'd want to feed it the shirt's size and color, and it outputs the product_variation's object

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
### WordPress Environment ###

Home URL: https://woocommerce.localhost
Site URL: https://woocommerce.localhost
WC Version: 3.4.1
Log Directory Writable: ✔
WP Version: ❌ 4.9.4 - There is a newer version of WordPress available (4.9.6)
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache
PHP Version: 7.2.1
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.52.1
OpenSSL/1.0.2j

SUHOSIN Installed: –
MySQL Version: 5.6.38
Max Upload Size: 8 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.4.1
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 3.36MB
Database Data Size: 2.11MB
Database Index Size: 1.25MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_bc_import_queue: Data: 0.02MB + Index: 0.05MB
wp_bc_products: Data: 0.02MB + Index: 0.09MB
wp_bc_variants: Data: 0.02MB + Index: 0.08MB
wp_commentmeta: Data: 0.06MB + Index: 0.03MB
wp_comments: Data: 0.02MB + Index: 0.09MB
wp_crosspeak_pending_updates: Data: 0.02MB + Index: 0.00MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_options: Data: 1.02MB + Index: 0.02MB
wp_postmeta: Data: 0.36MB + Index: 0.17MB
wp_posts: Data: 0.11MB + Index: 0.06MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_usermeta: Data: 0.02MB + Index: 0.03MB
wp_users: Data: 0.02MB + Index: 0.05MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB

### Post Type Counts ###

attachment: 103
bigcommerce_product: 14
nav_menu_item: 5
page: 8
post: 3
product: 41
product_variation: 8
revision: 13
shop_order: 3
wp_block: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (6) ###

CrossPeak OMS for WooCommerce: by CrossPeak OMS – 1.3.0 – Not tested with the active version of WooCommerce
Gutenberg Boilerplate: by Gary Thayer – 1.0.0
Gutenberg: by Gutenberg Team – 2.6.0
Woo Checkout Field Editor Pro: by ThemeHiGH – 1.2.5 – Not tested with the active version of WooCommerce
WooCommerce: by Automattic – 3.4.1
WordPress Importer: by wordpressdotorg – 0.6.3

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

### WC Pages ###

Shop base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My account: #7 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Sixteen
Version: 1.3 – 1.5 is available
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –
</details>

### PR DESCRIPTION
**Pull request by peterfabian**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #20349 .

### How to test the changes in this Pull Request:

1. Create attribute, e.g. color.
2. Create terms, e.g. yellow, blue, get ids of those terms.
3. Create shipping classes, e.g. `sc_1` and `sc_2`, get ids of those.
4. Create a tag, e.g. `tag_1`, get the id of the tag.
5. Create a category, e.g. `cat_1`, get the id of the category.
6.  Create a variable product with at least 3 variations: e.g. color = blue, color = yellow, color = unspecified
7. Set shipping class `sc_1` on the product level and assign `sc_2` to one of the variations.
8. Assign tag `tag_1` to the variable product.
9. Assign category `cat_1` to the variable product.
10. Set the product as featured.
11. Test `/wc/v2/products/{id}`, observe that `categories` and `tags` is empty.
12. Test `wc/v2/products/{id}/variations/?category=cat_1_id`, result is empty set.
13. Test `wc/v2/products/{id}/variations/?type=variable`, result is empty set.
14. Test `wc/v2/products/{id}/variations/?featured=true`, result is empty set. 
15. Test `wc/v2/products/{id}/variations/?featured=false`, it returns all variations, despite the product being featured.
16.  Test `wc/v2/products/{id}/variations/?tag=tag_1_id`, result is empty set.
17.  Test `wc/v2/products/{id}/variations/?shipping_class=sc_1_id`, result is empty set.
18.  Test `wc/v2/products/{id}/variations/?shipping_class=sc_2_id`, result is correct this time, as shipping class is assigned to the variation.
19. Test `wc/v2/products/{id}/variations/?attribute=pa_color&attribute_term=yellow_id`, result is empty set.
20. Apply branch
11. Test `/wc/v2/products/{id}`, observe that `categories` and `tags` is not empty anymore.
12. Test `wc/v2/products/{id}/variations/?category=cat_1_id`, result is array of all variations.
13. Test `wc/v2/products/{id}/variations/?type=variable`, result is array of all variations.
14. Test `wc/v2/products/{id}/variations/?featured=true`, result is array of all variations. 
15. Test `wc/v2/products/{id}/variations/?featured=false`, result is empty set, thus correct..
16.  Test `wc/v2/products/{id}/variations/?tag=tag_1_id`, result is array of all variations.
17.  Test `wc/v2/products/{id}/variations/?shipping_class=sc_1_id`, result is array of all variations that have shipping class 1 assigned to them.
18.  Test `wc/v2/products/{id}/variations/?shipping_class=sc_2_id`, result is array of all variations that have shipping class 1 assigned to them.
19. Test `wc/v2/products/{id}/variations/?attribute=pa_color&attribute_term=yellow_id`, result is array of all variations that have attribute color set to yellow.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### Changelog entry

> Allow filtering variations by attributes via query parameters in REST API.

### CODE REVIEW

File: includes/api/class-wc-rest-product-variations-controller.php [hunk 1]
```diff
@@ -39,6 +39,20 @@ class WC_REST_Product_Variations_Controller extends WC_REST_Products_Controller
 	 */
 	protected $post_type = 'product_variation';
 
+	/**
+	 * Name of parent product table in SQL query.
+	 *
+	 * @var string
+	 */
+	protected $parent_product_table_name = 'p_wc_variation_parent';
```
Comment (claudiosanches) [inline includes/api/class-wc-rest-product-variations-controller.php]: Seems like this is not used in any place. Probably just need to remove.

File: includes/wc-term-functions.php [hunk 2]
```diff
@@ -25,11 +25,11 @@ function wc_get_object_terms( $object_id, $taxonomy, $field = null, $index_key =
 	// Test if terms exists. get_the_terms() return false when it finds no terms.
 	$terms = get_the_terms( $object_id, $taxonomy );
 
-	if ( ! $terms || is_wp_error( $terms ) || is_null( $field ) ) {
+	if ( ! $terms || is_wp_error( $terms ) ) {
 		return array();
 	}
 
-	return wp_list_pluck( $terms, $field, $index_key );
+	return is_null( $field ) ? $terms : wp_list_pluck( $terms, $field, $index_key );
```
Comment (claudiosanches) [inline includes/wc-term-functions.php]: This changes will makes return different results now, suppose to return an empty string... Maybe include a new parameter or something else if need to return the list of terms even `$field` is `null`.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Just two smalls issues and then good to go.

File: includes/api/class-wc-rest-product-variations-controller.php (hunk 1 above)
Comment (peterfabian) [reply includes/api/class-wc-rest-product-variations-controller.php]: Good catch. Removed

File: includes/wc-term-functions.php (hunk 2 above)
Comment (peterfabian) [reply includes/wc-term-functions.php]: That was the point :) Please see https://github.com/woocommerce/woocommerce/pull/20079

Comment (claudiosanches) [reply includes/wc-term-functions.php]: @peterfabian ohh yeah, sorry.

**Review by claudiosanches — APPROVED** [review]

**Pull request merged**